### DISCUSSION TITLE
Improve generic type inference and nullability analysis with AllowNull

### DISCUSSION THREAD

**Discussion by mikernet** [dotnet/csharplang#3681] [Language Ideas]
The analyzer should be smart enough to see that all usages of `T1` in the parameters are allowed to be null and then infer `T1` to be `string` instead of `string?` so that the constraint is not violated:

```c#
string? value1 = null;
string value2 = null!;

public void Method<T1, T2>([AllowNull] T1 value, T2 value2) 
    where T1 : T2 
    where T2 : class?
{ }

Method(value1, value2); // string? is not a string so the contraint doesn't hold

// workaround:

Method<string, string>(value1, value2);
```

**Comment by mikernet** [comment]
[The related issue above](https://github.com/dotnet/roslyn/issues/39666) where I linked this issue was closed as "by design" but I would like to have a discussion about how in this particular case that's a bad design decision that should be eased up a bit. It should be possible for me to write methods in my library so that users of my library aren't getting nullability warnings when they are correctly using those methods.

**Comment by CyrusNajmabadi** [comment]
This looks like a language request FWIW.  Do you want me to move this to csharplang?  Also tagging @jcouv .

**Comment by mikernet** [comment]
Wasn't sure if nullability analysis belonged more in roslyn or csharplang, so if you feel it better fits there then go for it!

**Comment by mikernet** [comment]
I'd like to add a few notes from the gitter discussion around this issue to have here on record.

As per a comment @jnm2 made: if this was fixed to say "oh look, `[AllowNull] string` means `string?`", then it would infer `T1` to be `string` instead of `string?`

@CyrusNajmabadi expressed possible concerns about the consequences of the output values binding to a tighter type if `T1` was used as output. 

I can't think of any examples of that being a bad thing. The method body is already oblivious as to whether or not `T1` is actually nullable so that's not information it can act on, thus it already needs to consider that `T1` may not be nullable without knowing if it is.
